docs/performance-metrics.md

```diff
@@ -0,0 +1,98 @@
+# OpenShift ETCD Performance Metrics
+
+We often get complaints about ETCD performance being bad, this document should help track down the culprit and mitigate the issue at hand using Prometheus Metrics.
+
+## Common Bottlenecks
+
+### Memory
+
+ETCD uses bbolt, which stores everything as a single memory mapped file. In OpenShift this is up to 8 GiB big, this file is organized with a B+Tree and chunked into discrete 4k pages which can be flagged in-use or not. On top of that, several in-memory caches/indices on both ETCD and API server are being used to facilitate fast multi-version access and label/field selection respectively.
+
+Due to the indices and caches employed, it is fairly easy to exhaust memory on the control plane even though the actual used size of the etcd database file is tiny. One particular scenario where is this common is when listing all pods in a large cluster or listing many large configmaps/secrets.
+
+Check these metrics:
+
+> process_resident_memory_bytes{job="etcd"}
+> 
+> process_resident_memory_bytes{job="apiserver"}
+
+It is perfectly normal for etcd to use a multiple of its "in use DB size" as resident memory.
+
+Try to look through the audit log and find service accounts that might be listing pods or configmaps across namespaces.
+Another frequent offender are events from chatty operators, which can be found in the audit log or directly from [etcd using KCS6970015](https://access.redhat.com/solutions/6970015).
+
+#### Defrag and Compaction
+
+One way to reduce memory usage is to defrag bbolt and compact the database to the latest revision. It is important to run the defrag after the compaction, so one would execute:
+
+> etctctl compact $(etctctl get / -w json | jq .header.revision)
+> 
+> etctctl defrag
+
+In more recent OpenShift versions this is done automatically -  compacting from the API server and defrag from the etcd-operator.
```

Comment:
Is there a specific release you can ref here ie "in OpenShift 4.7+ this is done automatically"

Reply:
That would be Openshift 4.9+ I believe
https://github.com/openshift/cluster-etcd-operator/tree/release-4.9/pkg/operator/defragcontroller

Reply:
yep